### PR TITLE
Add split method to Strings in Rholang

### PR DESCRIPTION
**Pull request by Bill-Kunj**
## Overview

Add `explode` feature to strings for single-character delimiter string expansion
Closes RChip https://github.com/rchain/rchip-proposals/issues/52

### Notes

Does not include REGEX

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by tgrospic** [comment]
Thanks for contributing to RChain platform!

Please update you PR with corrections of these items:
1. Rebase PR to only include your changes. Two commits are now added which are not related to this PR.
1. Sign commits with GPG signature (link to more info is in the description of the PR).
1. Add relevant references to the PR description like RCHIP issues.

image

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 1]
```diff
@@ -1272,6 +1272,39 @@ class DebruijnInterpreter[M[_]: Sync: Parallel: _cost](
       } yield result
   }
 
+  private[this] val explode: Method = new Method() {
+
+    def explode(baseExpr: Expr, sep: Expr): M[Expr] =
```
Comment (tgrospic) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: @Bill-Kunj Please rename _explode_ to match Scala name `split`, in the code and in other places like PR title.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -1272,6 +1272,39 @@ class DebruijnInterpreter[M[_]: Sync: Parallel: _cost](
...
+      (baseExpr.exprInstance, sep.exprInstance) match {
+        case (GString(string), GString(sepStr)) =>
+          val split = string.split(sepStr)
```
Comment (tgrospic) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1280]: @Bill-Kunj Here, the problem is subtle and it's related how `F[_]` expressions are executed with monix so I'm not sure should we require you to fix it or maybe better for us to make final adjustment when all other data will be correct.

The problem is that splitting of string should happen (lazily) after `charge` is called. This means to put split inside F with the help of `Sync[F].delay` and not wrapping eagerly with `pure[F]`.
I'm OK if you leave this for us to adjust after all other data is corrected.

File: casper/src/main/scala/coop/rchain/casper/engine/LfsTupleSpaceRequester.scala [hunk 3]
```diff
@@ -35,9 +35,10 @@ object LfsTupleSpaceRequester {
     */
   type StatePartPath = Seq[(Blake2b256Hash, Option[Byte])]
 
-  // TODO: extract this as a parameter
-  // 20,000 records uses about 2G of direct buffer memory on importing side
-  val pageSize = 20000
+  /**
+    * Number of nodes in LFS sync data transfer
+    */
+  val pageSize = 750
```
Comment (dckc) [inline casper/src/main/scala/coop/rchain/casper/engine/LfsTupleSpaceRequester.scala:41]: is this change here on purpose? it seems unrelated.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 1 above)
Comment (Bill-Kunj) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Done

File: casper/src/main/scala/coop/rchain/casper/engine/LfsTupleSpaceRequester.scala (hunk 3 above)
Comment (Bill-Kunj) [reply casper/src/main/scala/coop/rchain/casper/engine/LfsTupleSpaceRequester.scala:41]: @TheoXD Can you rebase as suggested by @tgrospic? https://github.com/rchain/rchain/pull/3753#issuecomment-1162860649

Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/engine/LfsTupleSpaceRequester.scala:41]: We are not squashing commits on merge so intermediate changes on PR should be removed.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 4]
```diff
@@ -1272,6 +1272,39 @@ class DebruijnInterpreter[M[_]: Sync: Parallel: _cost](
       } yield result
   }
 
+  private[this] val split: Method = new Method() {
+
+    def split(baseExpr: Expr, sep: Expr): M[Expr] =
+      (baseExpr.exprInstance, sep.exprInstance) match {
+        case (GString(string), GString(sepStr)) =>
+          val split = string.split(sepStr)
+          charge[M](toListCost(split.size) + splitCost(string, sepStr, split.size)) >>
```
Comment (tgrospic) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1281]: @hilltracer Please verify if cost is correctly calculated?

Comment (hilltracer) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1281]: For split string with size 14:
“qwe1 qwe2 qwe3”.split(“ “) = RhoList(qwe1, qwe2, qwe3)

For me: 
size = 
14 - traverse each symbols of input string and compare with SepString
+
12 - Character-by-character copy of symbols from (qwe1qwe2qwe3) in array
+
toListCost(3) (now it’s equal 3, I think it’s wrong but it are used in many places - so doesn’t matter)

Conclusion:
```Cost = charge[M](toListCost(split.size) + splitCost(string, sepStr, split.size))```
where:  
```scala
def splitCost(str: String, sep: String, found: Int): Cost =
    Cost(2*str.getBytes.size - found * sep.length, “split”)
```
I think it’s a matter of taste.  You can ignore it, because yes...  it is probably possible to somehow optimize this and combine the traverse and copy operations, but then how to take into account the comparison with sep?

Comment (Bill-Kunj) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1281]: If this is a matter of taste, then let's use what's already coded? @tgrospic

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 2 above)
Comment (Bill-Kunj) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1280]: > @Bill-Kunj Here, the problem is subtle and it's related how `F[_]` expressions are executed with monix so I'm not sure should we require you to fix it or maybe better for us to make final adjustment when all other data will be correct.
> 
> The problem is that splitting of string should happen (lazily) after `charge` is called. This means to put split inside F with the help of `Sync[F].delay` and not wrapping eagerly with `pure[F]`. I'm OK if you leave this for us to adjust after all other data is corrected.

OK. We'll leave this as-is and hope to learn more

File: casper/src/main/scala/coop/rchain/casper/engine/LfsTupleSpaceRequester.scala (hunk 3 above)
Comment (Bill-Kunj) [reply casper/src/main/scala/coop/rchain/casper/engine/LfsTupleSpaceRequester.scala:41]: New PR submitted without intermediate changes: #3777

**Comment by Bill-Kunj** [comment]
Closed in lieu of #3777
